src/settings.vala

```diff
@@ -164,20 +164,31 @@ public class AGSettings : Object
             greeter.switch_contrast (value);
 
             var settings = Gtk.Settings.get_default ();
+            var theme_name = "";
+            var icon_theme_name = "";
+
             if (value)
             {
-                debug ("Switching GTK Theme to high contrast theme \"%s\"", AGSettings.get_string (AGSettings.KEY_HIGH_CONTRAST_THEME_NAME));
-                debug ("Switching icon theme to high contrast theme \"%s\"", AGSettings.get_string (AGSettings.KEY_HIGH_CONTRAST_ICON_THEME_NAME));
-                settings.set ("gtk-theme-name", AGSettings.get_string (AGSettings.KEY_HIGH_CONTRAST_THEME_NAME));
-                settings.set ("gtk-icon-theme-name", AGSettings.get_string (AGSettings.KEY_HIGH_CONTRAST_ICON_THEME_NAME));
+
+                /* FIXME: We need to check for wrong theme names here) and handle such flaws gracefully */
+
+                theme_name = AGSettings.get_string (AGSettings.KEY_HIGH_CONTRAST_THEME_NAME);
+                icon_theme_name = AGSettings.get_string (AGSettings.KEY_HIGH_CONTRAST_ICON_THEME_NAME);
+                debug ("Switching GTK Theme to high contrast theme \"%s\"", theme_name);
+                debug ("Switching icon theme to high contrast theme \"%s\"", icon_theme_name);
             }
             else
             {
-                debug ("Switching GTK Theme to default theme \"%s\"", this.default_theme_name_);
-                debug ("Switching icon theme to default icon theme \"%s\"", this.default_icon_theme_name_);
-                settings.set ("gtk-theme-name", this.default_theme_name_);
-                settings.set ("gtk-icon-theme-name", this.default_icon_theme_name_);
+
+                /* FIXME: We need to check for wrong theme names here) and handle such flaws gracefully */
+
+                theme_name = AGSettings.get_string (AGSettings.KEY_THEME_NAME);
+                icon_theme_name = AGSettings.get_string (AGSettings.KEY_ICON_THEME_NAME);
+                debug ("Switching GTK Theme to default theme \"%s\"", theme_name);
+                debug ("Switching icon theme to default icon theme \"%s\"", icon_theme_name);
             }
+            settings.set ("gtk-theme-name", theme_name);
+            settings.set ("gtk-icon-theme-name", icon_theme_name);
         }
     }
```

Comment:
Maybe... I wonder if we shouldn't use default_theme_name_ (resolving to Adwaita) as a fallback, in case someone puts a typo into KEY_THEME_NAME, similarly for default_icon_theme_name and KEY_ICON_THEME_NAME.

For highcontrast-mode, we could either hard-core "HighContrast" (and depend our packages on gnome-accessibility-themes) or also use Adwaita as fallback.

Reply:
I've looked up GTK's source code and it just doesn't make sense to make things more complicated than necessary here.

Instead, we should really just drop the `default_*_` variables and not query them in construct at all.

If invalid strings are provided, GTK will just do nothing and stay on the theme it currently has loaded.

In high contrast mode, we should certainly try to detect an invalid theme and just hardcode `HighContrast`, but there's no way that we could get this information from GTK. This makes our `default_*_` variables completely useless.

Reply:
Yes, here I concur. Go ahead and rip those `default_*_*` private properties out again.